James McConnell Type Modeling HW

jdmcconne opened the pull request:
Completed version of type modeling assignment + bonus problems.

pcantrell commented:
Good work!

For the final step of this assignment:

1. Open up the “Files changed” tab of this pull request.
2. Open up my own solution to the homework so that you can compare the two:
    - [Solution without bonus problems](https://github.com/mac-comp381/type-modeling/compare/solution-without-bonus...main)
    - [Solution with bonus problems](https://github.com/mac-comp381/type-modeling/compare/solution...main)

Walk through the two solutions, comparing one small chunk of code at a time. Things to think about as you go:

- Remember: different doesn’t necessarily mean wrong. Sometimes one way is clearly better than another, and sometimes you can just say, “Yup, that works too!”
- In the java-type-checker, did you ever rely on using strings to compare types? Would your solution work if there are two _different_ types that happen to have the same names?
- Are there differences between our approaches that would result in different language behavior — different type checking or different member lookup — but the directions and tests simply didn’t specify the desired behavior enough to make one clearly the correct one? (Think about the poor language implementers who have to deal with this sort of question all the time!)
- Are there Python or Java language features you notice that you’d like to investigate further?
- Are there Python or Java language features that _I_ should learn from you and investigate further? (This often happens with student work!)
- Did you forget to delete any commented-out or otherwise unused code, or extraneous debug statements?

Now, to complete the assignment:

**Write a brief note about one or two things that stood out to you as you compare the two solutions**. 

You can ask a question, or just name something notable. Post these comments **here in this comment thread** and/or as **comments on individual lines of your code** in the “Files changed” tab.

jdmcconne commented:
The or any feature in python is a very neat way to check for something in a list that I didn't know about.  The other major difference I found beyond not reducing duplicate code was the implementation of the MRO, but to be honest since that was one of the first things I added to the homework,  I don't think I had a great understanding of how the type system worked yet, and so now looking back it's hard for me to decode why I implemented it the way I did.  I found a lot of this homework to be me adding a few simple pieces of code and then running test expecting errors to fix, then they just passed and I'd have to go back and examine why the simple solution worked, which was a helpful process.  Another difference I noted was that I used is rather than == for a lot of my comparisons.  I did this to be stricter, but I'm not confident it's the best solution.

pcantrell commented:
The `is` vs `==` difference is an interesting one! Should ever allow, say, two different type objects that represent the same type? It depends on the context, how this stuff gets used. Your way is probably safer: it avoid accidentally introducing flexibility we might regret! My way _might_ be appropriate if we generate equivalent type objects in many different places during the AST-building process or something.
